app/ui/client/lib/UserAction.js

```diff
@@ -0,0 +1,177 @@
+import { ReactiveDict } from 'meteor/reactive-dict';
+import { Session } from 'meteor/session';
+import _ from 'underscore';
+
+import { settings } from '../../../settings';
+import { Notifications } from '../../../notifications';
+
+const timeout = 15000;
+const renew = timeout / 3;
+const rooms = {};
+
+export const USER_RECORDING = 'user-recording';
+export const USER_TYPING = 'user-typing';
+export const USER_UPLOADING = 'user-uploading';
+export const USER_ACTIVITY = 'user-activity';
+
+const recordingTimeouts = {};
+const uploadingTimeouts = {};
+const typingTimeouts = {};
+
+const recordingRenews = {};
+const uploadingRenews = {};
+const typingRenews = {};
+
+const recordingUsers = new ReactiveDict();
+const uploadingUsers = new ReactiveDict();
+const typingUsers = new ReactiveDict();
+
+const selfActivity = { USER_RECORDING: false, USER_UPLOADING: false, USER_TYPING: false };
+
+const shownName = function(user) {
+	if (!user) {
+		return;
+	}
+	if (settings.get('UI_Use_Real_Name')) {
+		return user.name;
+	}
+	return user.username;
+};
+
+const stopActivity = (rid, activityType, extras) => Notifications.notifyRoom(rid, USER_ACTIVITY, shownName(Meteor.user()), false, activityType, extras);
+const startActivity = (rid, activityType, extras) => Notifications.notifyRoom(rid, USER_ACTIVITY, shownName(Meteor.user()), true, activityType, extras);
+
+function handleStreamAction(activity, performingUsers, rid, username, extras) {
+	const id = extras?.tmid ? extras.tmid : rid;
+	const users = performingUsers.get(id) || {};
+
+	if (activity === true) {
+		clearTimeout(users[username]);
+		users[username] = setTimeout(function() {
+			const u = performingUsers.get(id);
+			delete u[username];
+			performingUsers.set(id, u);
+		}, timeout);
+	} else {
+		clearTimeout(users[username]);
+		delete users[username];
+	}
+
+	performingUsers.set(id, users);
+}
+
+export const UserAction = new class {
+	constructor() {
+		Tracker.autorun(() => Session.get('openedRoom') && this.addStream(Session.get('openedRoom')));
+	}
+
+	get selfActivity() { return selfActivity; }
+
+	addStream(rid) {
+		if (rooms[rid]) {
+			return;
+		}
+		rooms[rid] = function(username, activity, activityType, extras) {
+			const user = Meteor.users.findOne(Meteor.userId(), { fields: { name: 1, username: 1 } });
+			if (username === shownName(user)) {
+				return;
+			}
+			if (activityType === USER_RECORDING) {
+				handleStreamAction(activity, recordingUsers, rid, username, extras);
+			} else if (activityType === USER_UPLOADING) {
+				handleStreamAction(activity, uploadingUsers, rid, username, extras);
+			} else if (activityType === USER_TYPING) {
+				handleStreamAction(activity, typingUsers, rid, username, extras);
+			}
+		};
+		return Notifications.onRoom(rid, USER_ACTIVITY, rooms[rid]);
+	}
+
+	startPerformingAction(rid, activityType, timeouts, renews, extras) {
+		if (renews[rid]) {
+			return;
+		}
+		selfActivity[activityType] = true;
+
+		renews[rid] = setTimeout(() => {
```

Comment:
You can have single `renews` and `timeouts` here, just use `rid + activityType` as key instead of only `rid`. It will make the code simpler.